rfcs/0049-flakes.md

```diff
@@ -0,0 +1,708 @@
+  specified `rev` and/or `ref` are then merged with the entry in the
+  registry. (See below.) For example, `nixpkgs` and
+  `nixpkgs/release-19.03` are indirect flake references.
+
+In the future, we should also add tarball flake references
+(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
+straight-forward to add flake references to Mercurial repositories
+since Nix already has Mercurial support.
+
+## Flake registries
+
+Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
+"direct" flake references (i.e. any type of flake reference that's not
+an indirection). This is a convenience to users, allowing them to do
+
+    nix run nixpkgs#hello
+
+rather than
+
+    nix run github:NixOS/nixpkgs#hello
+
+There are multiple registries:
+
+* The global registry
+  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
+  location can be overriden via the `flake-registry` option.) Nix
+  automatically fetches this registry periodically. The check interval
+  is determined by the `tarball-ttl` option.
+
+* The local registry `$XDG_CONFIG_HOME/nix/registry.json`.
+
+* Registry entries specified on the command line.
+
+A registry is a JSON file that looks like this:
+
+```
+{
+    "version": 1,
+    "flakes": {
+        "nixpkgs": {
+            "uri": "github:NixOS/nixpkgs"
+        },
+        ...
+    }
+}
+```
+
+With this registry, flake references resolve as follows:
+
+* `nixpkgs` -> `github:NixOS/nixpkgs`
+* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
+* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`
+
+The registries are searched in reverse order. Thus the local registry
+overrides the global registry, and the command line takes precedence
+over the local and global registries.
+
+## `nix` command line interface
+
+I propose to make flakes the primary way to specify packages in the
+`nix` command line interface. Note that the `nix` UI is still marked
+as experimental, so we have some freedom to make incompatible
+changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
+`nix-instantiate`) should not be changed to avoid breakage.
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs#hello dwarffs#dwarffs
+
+takes two flake-based installables. The general form is:
+
+    <flake-ref>(#<attr-path>)?
+
+Examples of installables:
+
+* `nixpkgs#packages.hello`
+* `nixpkgs#hello` - short for `nixpkgs:packages.hello`
+* `nixpkgs/release-19.03#hello` - overrides the Git branch to use
+* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082#hello` -
+  specifies the exact Git revision to use
+* `dwarffs` - short for `dwarffs#defaultPackage`
+* `nix#hydraJobs.build.x86_64-darwin`
+* `.#hydraJobs.build.x86_64-darwin` - refers to the flake in the
```

Comment:
```.#``` here seems especially ugly. Some users may see examples like this and end up thinking this is a comment in your shell script.

Reply:
We could allow the dot to be left out, but then the hash would need to be quoted so it doesn't save any typing.